### PR TITLE
REF: implement _with_freq, use _from_sequence less

### PR DESCRIPTION
**Pull request by jbrockmendel**
This makes method-chaining a little bit nicer.  

Moving away from using _from_sequence and towards using the constructors where feasible.  This will make it easier to stricten what we accept in these _from_sequence methods.

We also should be inferring freq _less_ in arithmetic methods, but that is a behavior change that will be done separately.

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -396,6 +396,34 @@ def floor(self, freq, ambiguous="raise", nonexistent="raise"):
     def ceil(self, freq, ambiguous="raise", nonexistent="raise"):
         return self._round(freq, RoundTo.PLUS_INFTY, ambiguous, nonexistent)
 
+    def _with_freq(self, freq):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you type

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: Optional[Union[Dateoffset, str]]

Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: returns Union[DTA,TDI,PA] (but I think we have a type for this)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py]: just tried this and it looks like its going to turn into a yak-shaving event, prefer to do that separately

**Review by jreback — APPROVED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py]: kk, followup if you can

**Comment by jreback** [comment]
thanks

**Pull request merged**